README.md
# Ternary Decomposer
This was built for the Geological Survey of Brazil. 
The application takes as input a 16 million RGB geotiff file and decomposes it in 27 layers (check image files for references). The georeferencing of the output files is the same as the input. It also produces shapefiles (shp) of the tiff files.
 
## Dependencies (C++)
* Qt 5
* GDAL 3

## Dependencies (Python)
* GDAL/OSR/OGR (and libgdal-dev)
* Numpy
* PyQt

## Usage
Clone the repository and use `python main.py`.
The compiled version C++ is located at the Releases. When using the C++ version, you should open through the .bat file or files won't be georeferenced.

## Notes
* When importing to ArcGIS, it's necessary to disable some statistical calculation automatically does or the images will be badly displayed. This is solvable by simply disabling the ‘gamma stretch’ and the equalization to ‘none’ (properties tab of the raster). . 
* This application is expected to freeze for large files (however, just leave it open as it's doing the processing).
* You can bundle it through `pyinstaller -F main.py`.


Changelog.md
# Changelog

v0.2:
- C++: Bugfix - Included the .bat file for PROJ
- C++: Add PACKBITS compression for pszOptions

v0.1:
- Code written in Python
- Code written in C++

cpp/ternarydecomposer.cpp
#include "ternarydecomposer.h"
#include "qstring.h"
#include "qfile.h"
#include "gdal_priv.h"
#include <qmessagebox.h>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;
int vconversor(char rgb, uint16_t value);
bool colorchk(string comp, uint16_t k, uint16_t th, uint16_t u);

ternarydecomposer::ternarydecomposer(QWidget *parent) : QMainWindow(parent) {

	// Qt stuff
	QPushButton* fileSelector = new QPushButton("Select file and start!", this);
	fileSelector->resize(200, 30);
	fileSelector->move(64,60);
	connect(fileSelector, &QPushButton::clicked, this, &ternarydecomposer::flBtn_onClick);

	lbl = new QLabel("The processing will start after file selection.", this);
	lbl->resize(320, 60);
	lbl->move(20, 75);

	// TBD in future if necessary, GDAL documentation does not help at all
	//cbox = new QCheckBox("Produce shapefiles", this);
	//cbox->move(20, 60);
	//cbox->resize(320, 40);

	pBar = new QProgressBar(this);
	pBar->setGeometry(0, 0, 300, 25);
	pBar->setMaximum(100);
	pBar->move(50, 25);
}

void ternarydecomposer::flBtn_onClick() {
	int progress = 0;
	this->pBar->setValue(0);

	QString tiffFile = QFileDialog::getOpenFileName(this,
		tr("Select file"), "", tr("GeoTiff Files (*.tif)"));

	if (tiffFile.isEmpty()) {
		return;
	}
	else {
		// Sanity check 
		/*
		QFile file(tiffFile);
		if (!file.open(QIODevice::ReadOnly)) {
			QMessageBox::information(this, tr("Unable to open file"),
				file.errorString());
			return;
		}*/

		// Convert QString to const char
		QByteArray ba = tiffFile.toLocal8Bit();
		const char* fileLocation = ba.data();

		static vector<string> compositions;
		compositions = { "111", "112", "113", 
			"121", "122", "123",
			"131", "132", "133",
			"211", "212", "213",
			"221", "222", "223",
			"231", "232", "233",
			"311", "312", "313",
			"321", "322", "323",
			"331", "332", "333" };
		
		// Gdal stuff
		GDALAllRegister();
		GDALDataset* orig_ternary = (GDALDataset*)GDALOpen(fileLocation, GA_ReadOnly);
		const char* pszFormat = "GTiff";
		OGRSpatialReference oSRS;
		//oSRS = orig_ternary->GetSpatialRef();

		for (string composition : compositions) {
			this->pBar->setValue(progress);
			this->lbl->setText("Working...");
			GDALDataset* fused_bands;
			GDALDriver* tiffDriver;
			string filename = composition + ".tif";
			char k_pos = composition.at(0);
			char th_pos = composition.at(1);
			char u_pos = composition.at(2);

			// Get relevant data from original dataset
			int nRows = orig_ternary->GetRasterYSize();
			int nCols = orig_ternary->GetRasterXSize();
			
			//string wkt = proj = orig_ternary->GetProjectionRef();
			double noData = orig_ternary->GetRasterBand(1)->GetNoDataValue();
			
			//printf("Projection is `%s'\n", orig_ternary->GetProjectionRef());
			tiffDriver = GetGDALDriverManager()->GetDriverByName(pszFormat);

			char** papszOptions = NULL;
			papszOptions = CSLSetNameValue(papszOptions, "TILED", "NO");
			papszOptions = CSLSetNameValue(papszOptions, "COMPRESS", "PACKBITS");
			papszOptions = CSLSetNameValue(papszOptions, "PHOTOMETRIC", "RGB");
			papszOptions = CSLSetNameValue(papszOptions, "PROFILE", "GEOTIFF");
			//fused_bands = tiffDriver->CreateCopy(filename.c_str(), orig_ternary, FALSE, NULL, NULL, NULL);

			//fused_bands = tiffDriver->CreateCopy(filename.c_str(), orig_ternary, FALSE, papszOptions, GDALTermProgress, NULL);
			fused_bands = tiffDriver->Create(filename.c_str(), nCols, nRows, 3, GDT_Byte, papszOptions);
			GDALClose(fused_bands);
			fused_bands = (GDALDataset*)GDALOpen(filename.c_str(), GA_Update);
			fused_bands->GetRasterBand(1)->SetNoDataValue(0);
			fused_bands->GetRasterBand(2)->SetNoDataValue(0);
			fused_bands->GetRasterBand(3)->SetNoDataValue(0);

			// Allocate memory for row buffers
			uint16_t* k_Row = (uint16_t*)CPLMalloc(sizeof(uint16_t) * nCols);
			uint16_t* th_Row = (uint16_t*)CPLMalloc(sizeof(uint16_t) * nCols);
			uint16_t* u_Row = (uint16_t*)CPLMalloc(sizeof(uint16_t) * nCols);
			uint16_t* k_Row_new = (uint16_t*)CPLMalloc(sizeof(uint16_t) * nCols);
			uint16_t* th_Row_new = (uint16_t*)CPLMalloc(sizeof(uint16_t) * nCols);
			uint16_t* u_Row_new = (uint16_t*)CPLMalloc(sizeof(uint16_t) * nCols);

			// Decomposition
			for (auto i = 0; i < nRows; i++) {
				orig_ternary->GetRasterBand(1)->RasterIO(GF_Read, 0, i, nCols, 1, k_Row, nCols, 1, GDT_UInt16, 0, 0);
				orig_ternary->GetRasterBand(2)->RasterIO(GF_Read, 0, i, nCols, 1, th_Row, nCols, 1, GDT_UInt16, 0, 0);
				orig_ternary->GetRasterBand(3)->RasterIO(GF_Read, 0, i, nCols, 1, u_Row, nCols, 1, GDT_UInt16, 0, 0);
				for (auto j = 0; j < nCols; j++) {
					k_Row_new[j] = 0;
					u_Row_new[j] = 0;
					th_Row_new[j] = 0;

					// Sanity check
					if (colorchk(composition, vconversor(k_pos, k_Row[j]), vconversor(th_pos, th_Row[j]), vconversor(u_pos, u_Row[j]))) {
						k_Row_new[j] = vconversor(k_pos, k_Row[j]);;
						u_Row_new[j] = vconversor(u_pos, u_Row[j]);;
						th_Row_new[j] = vconversor(th_pos, th_Row[j]);
					}
				}
				GDALSetRasterColorInterpretation(fused_bands->GetRasterBand(1), GCI_RedBand);
				GDALSetRasterColorInterpretation(fused_bands->GetRasterBand(2), GCI_GreenBand);
				GDALSetRasterColorInterpretation(fused_bands->GetRasterBand(3), GCI_BlueBand);
				fused_bands->GetRasterBand(1)->RasterIO(GF_Write, 0, i, nCols, 1, k_Row_new, nCols, 1, GDT_UInt16, 0, 0);
				fused_bands->GetRasterBand(2)->RasterIO(GF_Write, 0, i, nCols, 1, th_Row_new, nCols, 1, GDT_UInt16, 0, 0);
				fused_bands->GetRasterBand(3)->RasterIO(GF_Write, 0, i, nCols, 1, u_Row_new, nCols, 1, GDT_UInt16, 0, 0);
				//fused_bands->SetProjection(orig_ternary->GetProjectionRef());
			}

			/* Will fix someday
			const char* pszDriverName = "ESRI Shapefile";
			string shpfilename = composition + ".shp";
			GDALDriver* shpDriver;
			GDALDataset* shpDS;
			OGRSpatialReference* shpproj = new OGRSpatialReference();
			shpproj->importFromWkt(proj);
			shpDriver = GetGDALDriverManager()->GetDriverByName(pszDriverName);
			shpDS = shpDriver->Create(shpfilename.c_str(), 0, 0, 0, GDT_Float32, NULL);
			OGRLayer* poLayer;
			poLayer = shpDS->CreateLayer(composition.c_str(), shpproj, wkbMultiPolygon, NULL);
			GDALFPolygonize(fused_bands->GetRasterBand(2), NULL, poLayer, -1, NULL, NULL, NULL);
			*/

			const char* pszProjection;
			pszProjection = orig_ternary->GetProjectionRef();
			double transform[6];
			const char* proj = orig_ternary->GetProjectionRef();
			orig_ternary->GetGeoTransform(transform);
			fused_bands->SetMetadata(orig_ternary->GetMetadata());
			if (!(transform[0] == 0 && transform[1] == 1 && transform[2] == 0 && transform[3] == 0 && transform[4] == 0 && transform[5] == 1)) {
				fused_bands->SetProjection(proj);
				fused_bands->SetGeoTransform(transform);
			}
			else if (orig_ternary->GetGCPCount() > 1) {
				fused_bands->SetGCPs(orig_ternary->GetGCPCount(), orig_ternary->GetGCPs(), orig_ternary->GetGCPProjection());
				
			}
			//fused_bands->SetSpatialRef(orig_ternary->GetSpatialRef());
			
			//fused_bands->SetGCPs(orig_ternary->GetGCPCount(), orig_ternary->GetGCPs(), orig_ternary->GetGCPSpatialRef());
			//fused_bands->FlushCache();
			GDALClose(fused_bands);
			progress = progress + 3.7;
			//GDALDestroyDriverManager(); // later

		}
		GDALClose(orig_ternary);
		this->pBar->setValue(100);
		QMessageBox processDone;
		processDone.setText("Processing is done!");
		processDone.exec();
		this->lbl->setText("The processing will start after file selection.");
		
	}

}


bool colorchk(string comp, uint16_t k, uint16_t th, uint16_t u) {
	
	int compvar = stoi(comp);
	if (compvar == 111 && k == 1 && th == 1 && u == 1) return true;
	else if (compvar == 112 && k == 1 && th == 1 && u == 128) return true;
	else if (compvar == 113 && k == 1 && th == 1 && u == 255) return true;
	else if (compvar == 121 && k == 1 && th == 128 && u == 1) return true;
	else if (compvar == 122 && k == 1 && th == 128 && u == 128) return true;
	else if (compvar == 123 && k == 1 && th == 128 && u == 255) return true;
	else if (compvar == 131 && k == 1 && th == 255 && u == 1) return true;
	else if (compvar == 132 && k == 1 && th == 255 && u == 128) return true;
	else if (compvar == 133 && k == 1 && th == 255 && u == 255) return true;
	else if (compvar == 211 && k == 128 && th == 1 && u == 1) return true;
	else if (compvar == 212 && k == 128 && th == 1 && u == 128) return true;
	else if (compvar == 213 && k == 128 && th == 1 && u == 255) return true;
	else if (compvar == 221 && k == 128 && th == 128 && u == 1) return true;
	else if (compvar == 222 && k == 128 && th == 128 && u == 128) return true;
	else if (compvar == 223 && k == 128 && th == 128 && u == 255) return true;
	else if (compvar == 231 && k == 128 && th == 255 && u == 1) return true;
	else if (compvar == 232 && k == 128 && th == 255 && u == 128) return true;
	else if (compvar == 233 && k == 128 && th == 255 && u == 255) return true;
	else if (compvar == 311 && k == 255 && th == 1 && u == 1) return true;
	else if (compvar == 312 && k == 255 && th == 1 && u == 128) return true;
	else if (compvar == 313 && k == 255 && th == 1 && u == 255) return true;
	else if (compvar == 321 && k == 255 && th == 128 && u == 1) return true;
	else if (compvar == 322 && k == 255 && th == 128 && u == 128) return true;
	else if (compvar == 323 && k == 255 && th == 128 && u == 255) return true;
	else if (compvar == 331 && k == 255 && th == 255 && u == 1) return true;
	else if (compvar == 332 && k == 255 && th == 255 && u == 128) return true;
	else if (compvar == 333 && k == 255 && th == 255 && u == 255) return true;
	return false;
}


int vconversor(char rgb, uint16_t value) {
	if (rgb == '1') {
		if (value <= 85) { return 1; }
		else { return 0; }
	}
	if (rgb == '2') {
		if (value > 85 && value < 170) { return 128; }
		else { return 0; }
	}
	if (rgb == '3') { 
		if (value >= 170) { return 255; }
		else { return 0; }
	}
}

main.py
import sys
import os
import ogr, osr
import gdal, gdalnumeric
import numpy as np
from PyQt5 import QtCore, QtWidgets
from PyQt5.QtWidgets import QMainWindow, QWidget, QLabel, QFileDialog, QProgressBar
from PyQt5.QtWidgets import QPushButton, QMessageBox, QCheckBox
from PyQt5.QtCore import QSize


# Ternary Decomposer
class MainWindow(QMainWindow):
    def __init__(self):
        QMainWindow.__init__(self)

        self.setMinimumSize(QSize(350, 220))    
        self.setWindowTitle("Ternary Decomposer") 
        
        # Main button
        fileBtn = QPushButton('Start decomposition!', self)
        fileBtn.clicked.connect(self.fileDialog)
        fileBtn.resize(200, 30)
        fileBtn.move(64, 100)

        cbox = QCheckBox("Produce Shapefiles", self)
        cbox.setChecked(False)
        cbox.move(20,60)
        cbox.resize(320,40)
        cbox.stateChanged.connect(self.clickBox)
        self.checkedBox = 0

        # Label
        self.processWarn = QLabel(self)
        self.processWarn.setWordWrap(True)
        self.processWarn.setText('<i>The processing will start automatically after the file selection.</i>')
        self.processWarn.resize(300,60)
        self.processWarn.move(20, 130)

        # PBar
        self.progressBar = QProgressBar(self)
        self.progressBar.setGeometry(0, 0, 300, 25)
        self.progressBar.setMaximum(100)
        self.progressBar.move(50, 25)

        # Process Status
        self.processStatus = QLabel(self)
        self.processStatus.setText('Waiting for file selection...')
        self.processStatus.resize(350,60)
        self.processStatus.move(90, 6)

    def clickBox(self, state):
        if state == QtCore.Qt.Checked:
            self.checkedBox = 1
            print(self.checkedBox)
        else:
            self.checkedBox = 0

    # File Dialog
    def fileDialog(self):
        options = QFileDialog.Options()
        fileName, _ = QFileDialog.getOpenFileName(self,"Ternary File", "","Geotiff Files (*.tif)", options=options)
        options |= QFileDialog.DontUseNativeDialog
        if fileName:
            self.raster2array(fileName)

    # Decompose
    def raster2array(self, filename):

        decomposition = [ '111', '112', '113', '121', '122', '123', '131', '132', '133', '211', '212', '213', '221', '222', '223', '231', '232', '233', '311', '312', '313', '321', '322', '323', '331', '332', '333']

        self.processStatus.setText('Loading bands...')
        self.progressBar.setValue(5)
        # Get file, open with GDAL and separate rasters
        fileDir = os.path.dirname(os.path.realpath('__file__'))
        source = gdal.Open(filename)
        K = np.array(source.GetRasterBand(1).ReadAsArray())
        Th = np.array(source.GetRasterBand(2).ReadAsArray())
        U = np.array(source.GetRasterBand(3).ReadAsArray())

        # Create zeros for all elements on low, mid, high
        K_ref = np.zeros(K.shape, dtype=np.uint8)
        Th_ref = np.zeros(Th.shape, dtype=np.uint8)
        U_ref = np.zeros(U.shape, dtype=np.uint8)

        self.processStatus.setText('Decomposing potassium...')
        self.progressBar.setValue(15)

        # Actual separation loop
        for i, j in np.ndindex(K.shape):
            if(K[i, j] <= 85):
                K_ref[i, j] = 1
            if(K[i, j] > 85 and K[i, j] <= 170):
                K_ref[i, j] = 128 
            if(K[i, j] > 170):
                K_ref[i, j] = 255        
        
        self.processStatus.setText('Decomposing thorium...')
        self.progressBar.setValue(25)

        
        # Actual separation loop
        for i, j in np.ndindex(Th.shape):
            if(Th[i, j] <= 85):
                Th_ref[i, j] = 1
            if(Th[i, j] > 85 and Th[i, j] <= 170):
                Th_ref[i, j] = 128 
            if(Th[i, j] > 170):
                Th_ref[i, j] = 255
        del Th

        self.processStatus.setText('Decomposing uranium...')
        self.progressBar.setValue(35)

        # Actual separation loop
        for i, j in np.ndindex(U.shape):
            if(U[i, j] <= 85):
                U_ref[i, j] = 1
            if(U[i, j] > 85 and U[i, j] <= 170):
                U_ref[i, j] = 128 
            if(U[i, j] > 170):
                U_ref[i, j] = 255        
        del U

        progressNumber = 1
        self.processStatus.setText('Exporting 1/27...')
    

        for i in decomposition:
            progressValue = 35 + 2.4
            outputName = filename + '_' + i  + '.tif'
            outputPath = os.path.join(fileDir, outputName)
            self.progressBar.setValue(progressValue)
            R = np.zeros(K.shape, dtype=np.uint8)
            G = np.zeros(K.shape, dtype=np.uint8)
            B = np.zeros(K.shape, dtype=np.uint8)
            progressText = 'Exporting' + str(progressNumber) + '/27...'

            for w, x in np.ndindex(K.shape):
                if(K_ref[w, x] == self.colortransform(i[0]) and Th_ref[w, x] == self.colortransform(i[1]) and U_ref[w, x] == self.colortransform(i[2])):
                    R[w, x] = self.colortransform(i[0])
                    G[w, x] = self.colortransform(i[1])
                    B[w, x] = self.colortransform(i[2])
            # Create Gtiff driver with RGB Opts   
            dst_ds = gdal.GetDriverByName('GTiff').Create(outputName, K.shape[1], K.shape[0], 3, gdal.GDT_Byte, options = ['PHOTOMETRIC=RGB', 'PROFILE=GeoTIFF',])
            # Copy georeferencing info
            gdalnumeric.CopyDatasetInfo(source, dst_ds)
            # Write arrays
            dst_ds.GetRasterBand(1).WriteArray(R)
            dst_ds.GetRasterBand(1).SetNoDataValue(0)
            dst_ds.FlushCache()
            dst_ds.GetRasterBand(2).WriteArray(G)
            dst_ds.GetRasterBand(2).SetNoDataValue(0)
            dst_ds.FlushCache()
            dst_ds.GetRasterBand(3).WriteArray(B)
            dst_ds.GetRasterBand(3).SetNoDataValue(0)
            dst_ds.FlushCache()

            if(self.checkedBox == 1):
                src_ds = dst_ds
                proj = osr.SpatialReference(wkt=src_ds.GetProjection())
                srcband = src_ds.GetRasterBand(2)
                dst_layername = i
                drv = ogr.GetDriverByName("ESRI Shapefile")
                dst_ds = drv.CreateDataSource( dst_layername + ".shp" )
                #gdalnumeric.CopyDatasetInfo(src_ds, dst_ds)
                dst_layer = dst_ds.CreateLayer(dst_layername, proj )
                gdal.Polygonize( srcband, None, dst_layer, -1, [], callback=None )
                dst_ds.FlushCache()
                
            dst_ds = None
        self.processStatus.setText('')
        self.progressBar.setValue(100)
        QMessageBox.about(self, "Ternary Decomposer", "It's done!")

    def colortransform(self, indexcolor):
        if (int(indexcolor) == 1):
            return 1
        if (int(indexcolor) == 2):
            return 128
        if (int(indexcolor) == 3):
            return 255

# Init Qt Window
if __name__ == "__main__":
    #appctxt = ApplicationContext()
    app = QtWidgets.QApplication(sys.argv)
    mainWin = MainWindow()
    mainWin.show()
    sys.exit(app.exec_())
